File: ssh_tunnel.py
"""Reverse SSH tunnel: forwards VDS:8088 -> localhost:8088"""
import sys
import time
import socket
import select
import threading
import paramiko

VDS_HOST = "80.87.203.161"
VDS_USER = "root"
VDS_PASS = "Dohod100mln$"
REMOTE_PORT = 8088       # port on VDS (nginx proxy_pass target)
LOCAL_HOST = "127.0.0.1"
LOCAL_PORT = 8088        # local dashboard

def handler(chan):
    sock = socket.socket()
    try:
        sock.connect((LOCAL_HOST, LOCAL_PORT))
    except Exception as e:
        print(f"[tunnel] forward connect failed: {e}")
        chan.close()
        return
    while True:
        r, _, _ = select.select([sock, chan], [], [], 60)
        if sock in r:
            data = sock.recv(32768)
            if not data:
                break
            chan.sendall(data)
        if chan in r:
            data = chan.recv(32768)
            if not data:
                break
            sock.sendall(data)
    chan.close()
    sock.close()

def reverse_tunnel(transport):
    transport.request_port_forward("127.0.0.1", REMOTE_PORT)
    print(f"[tunnel] Reverse tunnel active: VDS:{REMOTE_PORT} -> localhost:{LOCAL_PORT}")
    print(f"[tunnel] Dashboard URL: https://newtekpro.ru/ozon/")
    while True:
        chan = transport.accept(60)
        if chan is None:
            continue
        t = threading.Thread(target=handler, args=(chan,), daemon=True)
        t.start()

def main():
    while True:
        try:
            print(f"[tunnel] Connecting to {VDS_HOST}...")
            client = paramiko.SSHClient()
            client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            client.connect(VDS_HOST, username=VDS_USER, password=VDS_PASS, timeout=10)
            print(f"[tunnel] Connected!")
            reverse_tunnel(client.get_transport())
        except KeyboardInterrupt:
            print("[tunnel] Stopped by user")
            sys.exit(0)
        except Exception as e:
            print(f"[tunnel] Error: {e}, reconnecting in 5s...")
            time.sleep(5)

if __name__ == "__main__":
    main()

File: ssh_tunnel.js
const { Client } = require('ssh2');
const net = require('net');

const VDS_HOST = '80.87.203.161';
const VDS_USER = 'root';
const VDS_PASS = 'Dohod100mln$';
const REMOTE_PORT = 8088;
const LOCAL_HOST = '127.0.0.1';
const LOCAL_PORT = 8088;

function startTunnel() {
  const conn = new Client();

  conn.on('ready', () => {
    console.log('[tunnel] Connected to VDS');
    conn.forwardIn('0.0.0.0', REMOTE_PORT, (err) => {
      if (err) {
        console.error('[tunnel] forwardIn error:', err.message);
        conn.end();
        return;
      }
      console.log(`[tunnel] Reverse tunnel active: VDS:${REMOTE_PORT} -> localhost:${LOCAL_PORT}`);
      console.log('[tunnel] Dashboard URL: https://newtekpro.ru/ozon/');
    });
  });

  conn.on('tcp connection', (info, accept, reject) => {
    const channel = accept();
    const sock = net.createConnection(LOCAL_PORT, LOCAL_HOST, () => {
      channel.pipe(sock);
      sock.pipe(channel);
    });
    sock.on('error', (err) => {
      console.error('[tunnel] local connect error:', err.message);
      channel.close();
    });
    channel.on('error', () => sock.destroy());
    sock.on('close', () => channel.close());
    channel.on('close', () => sock.destroy());
  });

  conn.on('error', (err) => {
    console.error('[tunnel] SSH error:', err.message);
  });

  conn.on('close', () => {
    console.log('[tunnel] Connection closed, reconnecting in 5s...');
    setTimeout(startTunnel, 5000);
  });

  conn.on('keyboard-interactive', (_name, _instr, _lang, prompts, finish) => {
    finish(prompts.map(() => VDS_PASS));
  });

  conn.connect({
    host: VDS_HOST,
    username: VDS_USER,
    password: VDS_PASS,
    readyTimeout: 20000,
    keepaliveInterval: 15000,
    keepaliveCountMax: 3,
    tryKeyboard: true,
  });
}

startTunnel();
